src/back/msl/writer.rs

```diff
@@ -1265,9 +1213,376 @@ impl<W: Write> Writer<W> {
+                    }
+                    _ => {
+                        self.put_subscripted_access_chain(
+                            base,
+                            base_ty,
+                            index::GuardedIndex::Known(index),
+                            policy,
+                            context,
+                        )?;
+                    }
+                }
+            }
+            _ => self.put_expression(chain, context, false)?,
+        }
+
+        Ok(())
+    }
+
+    /// Write a `[]`-style access of `base` by `index`.
+    ///
+    /// If `policy` is [`Restrict`], then generate code as needed to force all index
+    /// values within bounds.
+    ///
+    /// The `base_ty` argument must be the type we are actually indexing, like [`Array`] or
+    /// [`Vector`]. In other words, it's `base`'s type with any surrounding [`Pointer`]
+    /// removed. Our callers often already have this handy.
+    ///
+    /// This only emits `[]` expressions; it doesn't handle struct member accesses or
+    /// referencing vector components by name.
+    ///
+    /// [`Restrict`]: crate::proc::index::BoundsCheckPolicy::Restrict
+    /// [`Array`]: crate::TypeInner::Array
+    /// [`Vector`]: crate::TypeInner::Vector
+    /// [`Pointer`]: crate::TypeInner::Pointer
+    fn put_subscripted_access_chain(
+        &mut self,
+        base: Handle<crate::Expression>,
+        base_ty: &crate::TypeInner,
+        index: index::GuardedIndex,
+        policy: index::BoundsCheckPolicy,
+        context: &ExpressionContext,
+    ) -> BackendResult {
+        let accessing_wrapped_array = match *base_ty {
+            crate::TypeInner::Array {
+                size: crate::ArraySize::Constant(_),
+                ..
+            } => true,
+            _ => false,
+        };
+
+        self.put_access_chain(base, policy, context)?;
+        if accessing_wrapped_array {
+            write!(self.out, ".{}", WRAPPED_ARRAY_FIELD)?;
+        }
+        write!(self.out, "[")?;
+
+        // Decide whether this index needs to be clamped to fall within range.
+        let restriction_needed = if policy == index::BoundsCheckPolicy::Restrict {
+            context.access_needs_check(base, index)
+        } else {
+            None
+        };
+        if let Some(limit) = restriction_needed {
+            write!(self.out, "{}::min(unsigned(", NAMESPACE)?;
+            self.put_index(index, context, true)?;
+            write!(self.out, "), ")?;
+            match limit {
+                index::IndexableLength::Known(limit) => {
+                    write!(self.out, "{}u)", limit - 1)?;
+                }
+                index::IndexableLength::Dynamic => {
+                    let global = base
+                        .originating_global(context.function)
+                        .ok_or(Error::Validation)?;
+                    self.put_dynamic_array_length(global, context)?;
+                    write!(self.out, " - 1)")?;
```

Comment:
this is a little unfortunate, since `put_dynamic_array_length` already adds "+1".
Perhaps, a better way to approach this would be to change the semantics of `put_dynamic_array_length` into `put_last_index_of_dynamic_array`? This would end up with the following changes:
  - the `arrayLength()` implementations would then append "+1", while checking for `is_scoped` to see if "("/")" are needed. This would be a slight improvement by itself!
  - the in-bounds checks could do "xxx <= last_index" instead of the current "xxx < last_index + 1"
  - the last index output would just be "last_index" instead of the current "last_index + 1 - 1"

Reply:
Great ideas. Done.